Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs

```diff
@@ -466,27 +466,47 @@ internal async Task<BoolResult> RestoreCheckpointAsync(OperationContext context,
 
         private async Task<BoolResult> ProcessStateAsync(OperationContext context, bool inline)
         {
-            try
+            var result = await processStateCoreAsync();
+
+            if (result.Succeeded &&
+                _postInitializationTask != null && _postInitializationTask.Status == TaskStatus.RanToCompletion && (await _postInitializationTask).Succeeded == false)
             {
-                var checkpointState = await GlobalStore.GetCheckpointStateAsync(context);
-                
-                if (!checkpointState)
-                {
-                    // The error is already logged.
-                    return checkpointState;
-                }
+                // A post initialization process may fail due to a transient issue, like a storage failure or an inconsistent checkpoint's state.
+                // The transient error can go away and the system may recover itself by calling this method again.
+
+                // In this case we need to reset _postInitializationTask and move it's state from "failure" to "success"
+                // and unblock all the public operations that will fail if post-initialization task is unsuccessful.
 
-                return await RestoreCheckpointAsync(context, checkpointState.Value, inline);
+                _postInitializationTask = Task.FromResult(result);
             }
-            finally
+
+            return result;
+
+            async Task<BoolResult> processStateCoreAsync()

```

Comment:
Minor: Not a fan of these private-methods-in-private-methods. Harder to distinguish inner method code from outer method code. In any case, needs to start with a comment describing function.

Reply:
I actually pretty like it. Because if you have two private methods: one is called ProcessStateAsync and another one is ProcessStateAndChangePostInitializationTaskAsync then it is possible to accidentally switch and use the wrong one.

In this case, the local method is intended to be local.

---
In reply to: [323499701](https://github.com/microsoft/BuildXL/pull/862#discussion_r323499701) [](ancestors = 323499701)